Move cuda driver exit handling from helpers to threads

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109739
* #110307
* #111072
* __->__ #111061

The pattern here is that main may exit and kill cuda driver before
c10d watchdog related threads have cleanly exited.  If this happens,
c10d threads may still make CUDA api calls and raise an exception about
the cuda driver being dead.

In the past we've patched a few helper functions that call into cuda
to specifically handle this driver exiting message.  Instead, we know
that this problem applies only to codepaths in our background threads,
so we should catch at that scope and not worry about fine-grained
catching at the helper granularity. (and if a helper is used from the main
thread, we should NOT catch this exception- it's the application's fault)

wconstab committed:
Move cuda driver exit handling from helpers to threads

The pattern here is that main may exit and kill cuda driver before
c10d watchdog related threads have cleanly exited.  If this happens,
c10d threads may still make CUDA api calls and raise an exception about
the cuda driver being dead.

In the past we've patched a few helper functions that call into cuda
to specifically handle this driver exiting message.  Instead, we know
that this problem applies only to codepaths in our background threads,
so we should catch at that scope and not worry about fine-grained
catching at the helper granularity. (and if a helper is used from the main
thread, we should NOT catch this exception- it's the application's fault)

[ghstack-poisoned]

wconstab committed:
Update on "Move cuda driver exit handling from helpers to threads"


The pattern here is that main may exit and kill cuda driver before
c10d watchdog related threads have cleanly exited.  If this happens,
c10d threads may still make CUDA api calls and raise an exception about
the cuda driver being dead.

In the past we've patched a few helper functions that call into cuda
to specifically handle this driver exiting message.  Instead, we know
that this problem applies only to codepaths in our background threads,
so we should catch at that scope and not worry about fine-grained
catching at the helper granularity. (and if a helper is used from the main
thread, we should NOT catch this exception- it's the application's fault)

[ghstack-poisoned]

malfet commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -922,15 +905,24 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
+          "] NCCL watchdog thread terminated with exception: ",
+          e.what());
+      LOG(ERROR) << exitMsg;
+      watchDogException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
+      std::rethrow_exception(watchDogException_);
```
I thought that `std::make_exception_ptr` is only needed/applicable if one wants to avoid downcasting type errors. I.e. what's wrong with good old throw?
Also, are there any docs on the lifetime of exception_ptr object? (I.e. I would not expect it to be a valid pointer outside of current stack unroll chain)
```suggestion
      throw watchDogException_ = std::runtime_error(exitMsg);
```

wconstab replied:
tagging @rohan-varma @H-Huang @kwen2501 for any context about this.  (we can probably address it in a later PR but it would be good to follow up)

fduwjj replied:
I don't understand that re-throw logic as well... Since this is a refactor, I think it's ok that we do it for now. We can probably add a TODO and I can follow up with the team later to see what is the best way to throw exceptions.

fduwjj reviewed (approved):
Agree with @malfet that we need to revisit the exception throwing logic and accept to unblock now.

wconstab committed:
Update on "Move cuda driver exit handling from helpers to threads"


The pattern here is that main may exit and kill cuda driver before
c10d watchdog related threads have cleanly exited.  If this happens,
c10d threads may still make CUDA api calls and raise an exception about
the cuda driver being dead.

In the past we've patched a few helper functions that call into cuda
to specifically handle this driver exiting message.  Instead, we know
that this problem applies only to codepaths in our background threads,
so we should catch at that scope and not worry about fine-grained
catching at the helper granularity. (and if a helper is used from the main
thread, we should NOT catch this exception- it's the application's fault)

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
